File: JavaScript/114.二叉树展开为链表.js
/*
 * @lc app=leetcode.cn id=114 lang=javascript
 *
 * [114] 二叉树展开为链表
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
/* 
思路：
就是递归，体会一下递归的魅力，相信flatten能做好它该做的（狗屁）
具体实现，让节点左边，等于null，让右边等于原先的左边，并且连接上之前右边的节点，然后递归各个节点。
*/
/* 
Accepted
225/225 cases passed (96 ms)
Your runtime beats 73.53 % of javascript submissions
Your memory usage beats 95.99 % of javascript submissions (39.7 MB)
*/
var flatten = function(root) {
  if (!root) return
  flatten(root.left)
  flatten(root.right)
  let left = root.left
  let right = root.right
  root.left = null
  root.right = left
  let p = root
  while(p.right) {
    p = p.right
  }
  p.right = right
};
// @lc code=end

File: Python3/114.二叉树展开为链表.py
#
# @lc app=leetcode.cn id=114 lang=python3
#
# [114] 二叉树展开为链表
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
"""
深度优先 后序遍历
Accepted
225/225 cases passed (28 ms)
Your runtime beats 96.59 % of python3 submissions
Your memory usage beats 71.47 % of python3 submissions (14.9 MB)
"""

"""
class Solution:
    def flatten(self, root: TreeNode) -> None:
        if not root:
            return
        self.flatten(root.left)
        self.flatten(root.right)

        right = root.right
        root.left, root.right = None, root.left
        while root.right:
            root = root.right
        root.right = right
"""

"""
深度优先，前序遍历
Accepted
225/225 cases passed (32 ms)
Your runtime beats 86.17 % of python3 submissions
Your memory usage beats 88.44 % of python3 submissions (14.8 MB)
"""

class Solution:
    def flatten(self, root: TreeNode) -> None:
        if not root:
            return
        stack = []

        def handle(node):
            if node:
                stack.append(node)
                handle(node.left)
                handle(node.right)

        handle(root)

        # while stack:
        #     prev = stack.pop(0)
        #     curr = None
        #     if stack:
        #         curr = stack.pop(0)
        #     prev.left = None
        #     prev.right = curr
        #     if curr:
        #         stack = [curr] + stack

        # 还是for循环写起来简洁一些
        for i in range(1, len(stack)):
            prev, curr = stack[i - 1], stack[i]
            prev.left = None
            prev.right = curr

# @lc code=end
